### PR TITLE
Fix highlights by chapter order 

### PR DESCRIPTION
**Pull request by janKozie1**
Is this approach fine? I will add the tests if it's okay.

for: https://app.zenhub.com/workspaces/openstax-unified-5b71aabe3815ff014b102258/issues/openstax/unified/743

### CODE REVIEW

File: src/app/content/highlights/components/Highlights.tsx
Comment (PiotrKozlowski) [inline src/app/content/highlights/components/Highlights.tsx]: Can you move this sorting logic to `summaryHighlights` selector `src/app/content/highlights/selectors.ts`?
Also: `const orderedPages = findTreePages(location)` should be probably before `sort` so it is not recalculated for every item.

**Review by PiotrKozlowski — CHANGES_REQUESTED** [review]

File: src/app/content/highlights/components/Highlights.tsx
Comment (PiotrKozlowski) [reply src/app/content/highlights/components/Highlights.tsx]: or actually create new selector based on `summaryHighlights` so it will be recalculated only if `summaryHighlights` has changed

Comment (PiotrKozlowski) [reply src/app/content/highlights/components/Highlights.tsx]: you can also replace `const locationFilters = useSelector(selectors.highlightLocationFilters);` with `const locationFilters = useSelector(selectors.highlightLocationFiltersWithContent);` so we will map only for locations with highlights

**Comment by TomWoodward** [comment]
@janKozie1 can you explain what the actual error was with safari?

**Comment by PiotrKozlowski** [comment]
Recording which @annaludwin posted seems not to be related to Safari
I think it might be related to this code: from `updateSummaryHighlightsDependOnFilters`
```
let newHighlights: SummaryHighlights = {
    ...summaryHighlights,
    [locationFilterId]: {
      ...summaryHighlights[locationFilterId],
      [pageId]: [...(summaryHighlights[locationFilterId] || {})[pageId] || []],
    },
  };
```
we are adding `[pageId]` at the and of object which then is rendered also at the end?

This code will be removed with https://github.com/openstax/rex-web/pull/478 so this actually will not need any more work?

**Comment by janKozie1** [comment]
I was able to replicate  in Calculus Volume 1 - functions and graphs by:
- adding a highlight in 1.3
- adding a highlight in 1.5
- adding a highlight in 1.1

File: src/app/content/highlights/selectors.ts
Comment (TomWoodward) [inline src/app/content/highlights/selectors.ts]: if what you want is `highlights` you should depend on `summaryHighlights` not `localState` so the memoization works better

File: src/app/content/highlights/utils/summaryHighlightsUtils.ts
Comment (TomWoodward) [inline src/app/content/highlights/utils/summaryHighlightsUtils.ts]:
```suggestion
export const getSortedSummaryHighlights =
```

File: src/app/content/highlights/utils/summaryHighlightsUtils.ts
Comment (TomWoodward) [inline src/app/content/highlights/utils/summaryHighlightsUtils.ts]: does this function return `SummaryHighlights`? objects have no implicit order in javascript, so if this is supposed to sort things we should really be returning an array

**Review by TomWoodward — CHANGES_REQUESTED** [review]

File: src/app/content/highlights/selectors.ts
Comment (PiotrKozlowski) [inline src/app/content/highlights/selectors.ts]: It seems like this change is unnecessary

File: src/app/content/highlights/utils/summaryHighlightsUtils.ts
Comment (PiotrKozlowski) [inline src/app/content/highlights/utils/summaryHighlightsUtils.ts]: Can we simplify this to something like: ?
```
const pages: OrderedSummaryHighlights[0]['pages'] = [];

      for (const page of orderedPages) {
        if (!summaryHighlights[locationID][page.id]) { continue; }
        pages.push({
          highlights: summaryHighlights[locationID][page.id],
          pageId: page.id,
        });
      }

      return [
        ...previousLocations,
        {
          location,
          pages,
        },
      ];
```

**Review by PiotrKozlowski — CHANGES_REQUESTED** [review]
It looks good, I've left two comments, but I think you can start adding tests

**Comment by janKozie1** [comment]
I don't think the test for `SectionHighlights` to throw is necessary anymore since the pages are based on `findTreePages(location)`

**Review by PiotrKozlowski — APPROVED** [review]

File: src/app/content/highlights/components/Highlights.tsx
Comment (TomWoodward) [inline src/app/content/highlights/components/Highlights.tsx]: this line break doesn't really have to be here

**Review by TomWoodward — CHANGES_REQUESTED** [review]

**Comment by janKozie1** [comment]
@TomWoodward updated

**Review by TomWoodward — APPROVED** [review]
wait for qa review before merging

**Pull request merged**
